### PR TITLE
fix: allow absolute classpath

### PR DESCRIPTION
**Pull request by davecramer**
The current behaviour is that classpath is always relative
to $GPHOME/lib/postgresql/java/

The implication of this is that in order to reference jars outside
of this directory the classpath will look like ../../../ etc.

This patch retains that behaviour but adds the ability to use absolute
paths.

Note: without changing the security policy for the JVM this will only
work with the untrusted language javau. Otherwise Security Exceptions
will be thrown.

### CODE REVIEW

**Comment by ivannovick** [comment]
Hi Dave when this is merge, can you please also help to backpatch this to 4_3 stable as its considered a customer bug fix, we want to get it to our customers asap.

**Comment by ivannovick** [comment]
Hi Dave, can you also please comment with the information required for the documentation to be updated, i can ask someone else to update the docs, but need to understand how the behavior is changed for users.  Thx!

**Comment by davecramer** [comment]
This change retains the previous behaviour of requiring the classpath to be
relative to $GPHOME/lib/postgresql/java/

It adds the ability for the untrusted version of the language to add
classpaths which are absolute.

In order for this to work the classpath has to start with / for example

set pljava_classpath='/usr/local/gpdb/share/postgresql/pljava/examples.jar';

Dave Cramer

**Comment by ivannovick** [comment]
Hi Dave,

I talked to a customer waiting for this fix today and got the following feedback.

We are both confused about the untrusted part. 

The customer wants to use the version of PL/Java which is safer and allows for application developers that are NOT superuser to create and run functions.  The other version which is unsafe is only available for superusers and not the one that is going to be used in large commercial environment.

So the safe version is "trusted" or "untrusted" ?  Sorry its so confusing.

I want to make sure the fix will work for our customer that is using in the "safe" and recommended way for large commercial environment, does your fix support that?

**Comment by davecramer** [comment]
Hi Ivan,

untrusted means there are no restrictions on what the language can do, and
you need to be superuser to create it.

From the docs:

TRUSTED specifies that the language does not grant access to data that the
user would not otherwise have. If this key word is omitted when registering
the language, only users with the PostgreSQL superuser privilege can use
this language to create new functions.

As to my patch. it works with the version of PL/Java which is safer.
However that is not part of my patch but rather part of the java security
policy in general. If you attempt to use pl/java with my patch and specify
a CLASSPATH outside of the lib dir you will get a security exception from
the JVM.

The short version is yes. However they will have to do one of two things:

1) change all function defitnitions to use create function .... language
javau
or
2) change the install sql to CREATE LANGUAGE java HANDLER
pljavau_call_handler from CREATE LANGUAGE javaU HANDLER
pljavau_call_handler;

I would see 2 as the right option so they don't have to change all of their
SQL. I would just not install the trusted version.

yes, the trusted/untrusted is a bit confusing.

Dave Cramer

**Comment by ivannovick** [comment]
Hi Dave,

I am still confused, sorry.  So it sounds like the customer wants the TRUSTED version which will allow them to use it as non-super user.  The reasons are they want the extra security protections provided by the TRUSTED version and by the JVM.  Did i get this right?

So it sounds like we are blocked because with TRUSTED the JVM will not allow the path to be a random location for class files it needs to be inside the lib dir?

So for example /nfs/mount1/application1/lib/myclasses.class

The user wants to store his java class files and have them loaded.  Will the TRUSTED version be able to support this before or after the patch?  Right now it works for them, but they use ../../../../nfs/mount1/application1/lib/myclasses.class i believe to make it work.

**Comment by davecramer** [comment]
Hi Ivan,

OK, so the next question is how do we specify the paths that the JVM will
allow. That may have to go into a GUC. Can assume that is secure?

as to your last question. Both absolute and relative paths now work for
loading class files.
Dave

Dave Cramer

**Comment by schubert** [comment]
Ah well clearly the bot needs correcting here. I'll take care of it. Sorry for the noise on this PR.

**Comment by schubert** [comment]
Actually we've double checked the bot and we think it's fine, this might be a case of organization membership visibility.

@davecramer you'll want to adjust your organization visibility settings per the bot's instructions above.

**Comment by davecramer** [comment]
@ivannovick I've confirmed that the patch allows loading jars into the classpath from anywhere on the filesystem.

The trusted version of the language will enforce the JVM security policy which does not allow writes as you surmised.

The untrusted version does not enforce the JVM security policy.

File: gpAux/extensions/pljava/src/java/pljava/org/postgresql/pljava/sqlj/Loader.java [hunk 1]
```diff
@@ -95,7 +95,20 @@
 		{
 			try 
 			{
-				URL url = new URL("file:///" + jarpath + "/" + m_classpath[i]);
+				String path="file:///";
+				
+				// Previous behaviour required that paths be relative to $GPHOME/lib/postgresql/java/
```
Comment (davecramer) [inline gpAux/extensions/pljava/src/java/pljava/org/postgresql/pljava/sqlj/Loader.java]: @hlinnaka is this acceptable ?

Comment (hlinnaka) [inline gpAux/extensions/pljava/src/java/pljava/org/postgresql/pljava/sqlj/Loader.java]: Yeah, thta's better

**Comment by ivannovick** [comment]
Hi guys,

I got more input from users of this functionality yesterday. 

Here is the requirement they gave:

1) Use the "fenced"/"safe" version (sorry i still can't remember if that is trusted/untrusted)
2) Have a GUC to set the class path at system level by super users 
3) Path is not required by end users when specifying classes, because they are in the class path
4) If the classes are not in the class path end-users can not use the java class files

Does our existing patch meet this requirement?  Thx!

Ivan

**Comment by davecramer** [comment]
Yes, this is the functionality today, except for the semantics of the GUC.

Currently any user can specify the path. The only way the path can be set
at a system level is to set it in postgresql.conf

I will update the patch to fix this so that only super users can change the
path which will mean it will have to be set in postgresql.conf

Dave Cramer

**Comment by davecramer** [comment]
Ivan,

Making this change will break existing apps.

They will no longer be able to specify the classpath in their sql.

Dave Cramer
